nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -65,44 +63,78 @@ impl DataStore {
+    /// query can theoretically fail if someone is able to delete that pool or
+    /// make another one the default and delete that.
+    pub async fn ip_pools_fetch_default(
         &self,
         opctx: &OpContext,
-        action: authz::Action,
-    ) -> LookupResult<(authz::IpPool, IpPool)> {
-        self.ip_pools_fetch_for(
-            opctx,
-            action,
-            &Name(ExternalName::from_str("default").unwrap()),
-        )
-        .await
+    ) -> LookupResult<IpPool> {
+        use db::schema::ip_pool::dsl;
+
+        let authz_silo_id = opctx.authn.silo_required()?.id();
+
+        // TODO: Need auth check here. Only fleet viewers can list children on
+        // IP_POOL_LIST, so if we check that, nobody can make instances. This
+        // used to check CreateChild on an individual IP pool, but now we're not
+        // looking up by name so the check is more complicated
+        //
+        // opctx
+        //     .authorize(authz::Action::ListChildren, &authz::IP_POOL_LIST)
+        //     .await?;
+
+        dsl::ip_pool
+            .filter(dsl::silo_id.eq(authz_silo_id).or(dsl::silo_id.is_null()))
+            .filter(dsl::is_default.eq(true))
+            .filter(dsl::time_deleted.is_null())
+            // this will sort by most specific first, i.e.,
+            //
+            //   (silo)
+            //   (null)
+            //
+            // then by only taking the first result, we get the most specific one
+            .order(dsl::silo_id.asc().nulls_last())
+            .select(IpPool::as_select())
+            .first_async::<IpPool>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
-    /// Looks up an IP pool by name.
-    pub(crate) async fn ip_pools_fetch_for(
+    /// Looks up an IP pool by name if it does not conflict with your current scope.
+    pub(crate) async fn ip_pools_fetch(
         &self,
         opctx: &OpContext,
-        action: authz::Action,
         name: &Name,
-    ) -> LookupResult<(authz::IpPool, IpPool)> {
+    ) -> LookupResult<IpPool> {
+        let authz_silo_id = opctx.authn.silo_required()?.id();
+
         let (.., authz_pool, pool) = LookupPath::new(opctx, &self)
             .ip_pool_name(&name)
-            .fetch_for(action)
+            // any authenticated user can CreateChild on an IP pool. this is
+            // meant to represent allocating an IP
+            .fetch_for(authz::Action::CreateChild)
             .await?;
-        if pool.internal {
-            return Err(authz_pool.not_found());
+
+        // You can't look up a pool by name if it conflicts with your current
+        // scope, i.e., if it has a silo it is different from your current silo
+        if let Some(pool_silo_id) = pool.silo_id {
+            if pool_silo_id != authz_silo_id {
+                return Err(authz_pool.not_found());
+            }
```

Comment:
```suggestion
        if Some(authz_silo_id) != pool.silo_id {
            return Err(authz_pool.not_found());
```

Reply:
This doesn't work when `pool.silo_id == None`. We only want to do the check if the fetched pool has a silo ID. If `pool.silo_id.is_none()`, then it can't conflict with the current silo because it's a fleet-wide pool.